src/main/scala-3/com/raquo/airstream/split/SplittableTypeMacros.scala

```diff
@@ -0,0 +1,417 @@
+  * enum Baz extends Foo {
+  *   case Baz1, Baz2
+  * }
+  * case object Tar extends Foo
+  * val splitter = fooSignal.splitMatch
+  *   .handleCase { case Bar(Some(str)) => str } { (str, strSignal) =>
+  *     renderStrNode(str, strSignal)
+  *   }
+  *   .handleCase { case baz: Baz => baz } { (baz, bazSignal) =>
+  *     renderBazNode(baz, bazSignal)
+  *   }
+  *   .handleCase {
+  *     case Tar    => ()
+  *     case _: Int => ()
+  *   } { (_, _) => div("Taz") }
+  *   .toSignal
+  * ```
+  *
+  * into this code:
+  *
+  * ```scala
+  * val splitter = fooSignal.
+  *  .map { i =>
+  *    i match {
+  *      case Bar(Some(str)) => (0, str)
+  *      case baz: Baz => (1, baz)
+  *      case Tar => (2, ())
+  *      case _: Int => (2, ())
+  *    }
+  *  }
+  *  .splitOne(_._1) { ??? }
+  * ```
+  *
+  * After macros expansion, compiler will warns above code "match may not be
+  * exhaustive" and "unreachable case" as expected.
+  */
+object SplittableTypeMacros {
+
+  extension [Self[+_] <: Observable[_], I](inline observable: BaseObservable[Self, I]) {
+    inline def splitMatch: MatchSplitObservable[Self, I, Nothing] =
+      MatchSplitObservable.build(
+        observable,
+        Nil,
+        Map.empty[Int, Function2[Any, Any, Nothing]]
+      )
+  }
+
+  extension [Self[+_] <: Observable[_], I, O](
+    inline matchSplitObservable: MatchSplitObservable[Self, I, O]
+  ) {
+    inline def handleCase[A, B, O1 >: O](inline casePf: PartialFunction[A, B])(inline handleFn: (B, Signal[B]) => O1) = ${
+      handleCaseImpl('{ matchSplitObservable }, '{ casePf }, '{ handleFn })
+    }
+
+    inline private def handlePfType[T](inline casePf: PartialFunction[Any, T]) = ${
+      handleTypeImpl[Self, I, O, T]('{ matchSplitObservable }, '{ casePf })
+    }
+
+    inline def handleType[T]: MatchTypeObservable[Self, I, O, T] = handlePfType[T] { case t: T => t }
+
+    inline private def handlePfValue[V](inline casePf: PartialFunction[Any, V]) = ${
+      handleValueImpl[Self, I, O, V]('{ matchSplitObservable }, '{ casePf })
+    }
+
+    inline def handleValue[V](inline v: V)(using inline valueOf: ValueOf[V]): MatchValueObservable[Self, I, O, V, V] = handlePfValue[V] { case _: V => v }
+  }
+
+  extension [Self[+_] <: Observable[_], I, O, T](inline matchTypeObserver: MatchTypeObservable[Self, I, O, T]) {
+    inline def apply[O1 >: O](inline handleFn: (T, Signal[T]) => O1): MatchSplitObservable[Self, I, O1] = ${
+      handleTypeApplyImpl('{ matchTypeObserver }, '{ handleFn })
+    }
+  }
+
+  extension [Self[+_] <: Observable[_], I, O, V0, V1](inline matchValueObservable: MatchValueObservable[Self, I, O, V0, V1]) {
+    inline private def deglate[O1 >: O](inline handleFn: (V1, Signal[V1]) => O1) = ${
+      handleValueApplyImpl('{ matchValueObservable }, '{ handleFn })
+    }
+
+    inline def apply[O1 >: O](inline handleFn: Signal[V1] => O1): MatchSplitObservable[Self, I, O1] = deglate { (_, vSignal) => handleFn(vSignal) }
```

Comment:
difference in signature (`Signal[V1] => O1` here) means we need another (quite redundant IMHO) dummy class and dummy "detour".